neonvm-runner/cmd/main.go

```diff
@@ -669,6 +673,91 @@ func main() {
 	}
 }
 
+type NetworkMonitoringMetrics struct {
+	IngressBytes, EgressBytes       prometheus.Counter
+	Errors                          *prometheus.CounterVec
+	IngressBytesRaw, EgressBytesRaw uint64 // Absolute values to calc increments for Counters
+}
+
+func NewMonitoringMetrics(reg *prometheus.Registry) *NetworkMonitoringMetrics {
+	m := &NetworkMonitoringMetrics{
+		IngressBytes: util.RegisterMetric(reg, prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "runner_vm_ingress_bytes",
+				Help: "Number of bytes received by the VM from the open internet",
+			},
+		)),
+		EgressBytes: util.RegisterMetric(reg, prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "runner_vm_egress_bytes",
+				Help: "Number of bytes sent by the VM to the open internet",
+			},
+		)),
+		IngressBytesRaw: 0,
+		EgressBytesRaw:  0,
+		Errors: util.RegisterMetric(reg, prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "runner_vm_network_fetch_errors_total",
+				Help: "Number of errors while fetching network monitoring data",
+			},
+			[]string{"cause"},
+		)),
+	}
+	return m
+}
+
+func getNetworkBytesCounter(iptables *iptables.IPTables, chain string) (uint64, error) {
+	cnt := uint64(0)
+	rules, err := iptables.Stats("filter", chain)
+	if err != nil {
+		return cnt, err
+	}
+
+	// We need to measure only external traffic to/from vm, so we filter internal traffic
+	for _, rawStat := range rules {
+		stat, err := iptables.ParseStat(rawStat)
+		if err != nil {
+			return cnt, err
+		}
+		if stat.Protocol != "6" { // count tcp only
+			continue
+		}
+		src, dest := stat.Source.IP, stat.Destination.IP
+		if src.IsUnspecified() || dest.IsUnspecified() ||
+			src.IsLoopback() || dest.IsLoopback() ||
+			src.IsPrivate() || dest.IsPrivate() {
```

Comment:
Lots of duplication here, can we extract the `!ip.IsUnspecified() && !ip.IsLoopback() && !ip.IsPrivate()` into some shared function?